### PR TITLE
Adding the provider

### PR DESCRIPTION
**Pull request by fabi200123**
- Adding the GCP External Provider
- Update README.md

### CODE REVIEW

File: internal/client/gcp.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+	}
+
+	if err = op.Wait(ctx); err != nil {
+		return fmt.Errorf("unable to wait for the operation: %w", err)
+	}
+
+	return nil
+}
+
+func selectStartupScript(osType string) string {
```
Comment (gabriel-samfira) [inline internal/client/gcp.go:238]: you can use `params.OSType` (https://github.com/cloudbase/garm-provider-common/blob/main/params/params.go#L28-L32) here, and skip casting to string.

File: internal/spec/spec.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		udata, err := cloudconfig.GetRunnerInstallScript(r.BootstrapParams, r.Tools, r.BootstrapParams.Name)
+		if err != nil {
+			return "", fmt.Errorf("failed to generate userdata: %w", err)
+		}
+
+		lines := strings.Split(string(udata), "\n")
+		if len(lines) > 0 && strings.HasPrefix(lines[0], "#!") {
+			additionalCommands := []string{
+				// Create user 'runner' if it doesn't exist; '|| true' to ignore if user already exists
+				"sudo useradd -m " + defaults.DefaultUser + " || true",
```
Comment (gabriel-samfira) [inline internal/spec/spec.go:139]: You also need to make sure the new user is a part of the sudo/admin group.

File: internal/util/util.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		OSArch:     params.OSArch(*gcpInstance.Disks[0].Architecture),
+	}
+
+	switch gcpInstance.GetStatus() {
+	case "RUNNING":
+		details.Status = params.InstanceRunning
+	case "STOPPING", "TERMINATED", "SUSPENDED":
+		details.Status = params.InstanceStopped
+	case "PROVISIONING":
+		details.Status = params.InstancePendingCreate
```
Comment (gabriel-samfira) [inline internal/util/util.go:53]: Careful, GARM validates the runner status. Providers may only return one of the following statuses:

https://github.com/cloudbase/garm/blob/7c44b8125a33dd903cb1d275eb5eaf580fdbe1d7/runner/providers/external/util.go#L12-L21

File: provider/provider.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+func (g *GcpProvider) CreateInstance(ctx context.Context, bootstrapParams params.BootstrapInstance) (params.ProviderInstance, error) {
+	spec, err := spec.GetRunnerSpecFromBootstrapParams(g.cfg, bootstrapParams, g.controllerID)
+	if err != nil {
+		return params.ProviderInstance{}, fmt.Errorf("failed to get runner spec: %w", err)
+	}
+	inst, err := g.gcpCli.CreateInstance(ctx, spec)
+	if err != nil {
+		return params.ProviderInstance{}, fmt.Errorf("error creating instance: %w", err)
+	}
+	instance := params.ProviderInstance{
```
Comment (gabriel-samfira) [inline provider/provider.go:65]: You can theoretically use: `GcpInstanceToParamsInstance(inst)` here

**Review by gabriel-samfira — CHANGES_REQUESTED** [review]
Just a few small comments. I have some doubts in regards to the startup script approach on Linux.

* The runner user needs to be part of the "admin" group and be able to use passwordless sudo
* The sudo user must exist
* curl and tar must be installed

Normally, the package management and user creation management is done by cloud-init, but if we can;t leverage it here, we need to make sure we have a robust enough startup script that it can take care of all the dependency on our own.

File: internal/spec/spec.go (hunk 2 above)
Comment (fabi200123) [reply internal/spec/spec.go:139]: Added the commands to:
- add the `runner` user to the sudo group
- install curl and tar

File: internal/util/util.go (hunk 3 above)
Comment (fabi200123) [reply internal/util/util.go]: Updated it! Thanks!

File: internal/spec/spec.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+				// Create the runner home directory if it doesn't exist
+				"sudo mkdir -p /home/" + defaults.DefaultUser,
```
Comment (gabriel-samfira) [inline internal/spec/spec.go:141]: You also need to `chown runner:runner -R /home/runner`.

**Review by gabriel-samfira — CHANGES_REQUESTED** [review]
One last change. I am still not confident in the init script, but let's see what happens.

File: internal/spec/spec.go (hunk 5 above)
Comment (fabi200123) [reply internal/spec/spec.go:141]: This is done already in the generated script:
```
sendStatus "enabling runner service"
cp /home/runner/actions-runner/bin/runsvc.sh /home/runner/actions-runner/ || fail "failed to copy runsvc.sh"
sudo chown runner:runner -R /home/runner || fail "failed to change owner"
sudo systemctl daemon-reload || fail "failed to reload systemd"
sudo systemctl enable $SVC_NAME
```

Comment (gabriel-samfira) [reply internal/spec/spec.go:141]: Good point

**Pull request merged**